### PR TITLE
Preserve rack FailureDomain if previously assigned

### PR DESCRIPTION
**Pull request by egafford**
This change ensures that if a cluster has already been assigned the
rack failure domain in previous reconcile loops, it will continue
to use that failure domain.

Signed-off-by: Elise Gafford <egafford@redhat.com>

### CODE REVIEW

File: pkg/controller/storagecluster/reconcile.go [hunk 1]
```diff
@@ -371,12 +371,15 @@ func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageClus
...
-					}
+			if strings.Contains(label, "rack") {
+				if !nodeRacks.Contains(value, node.Name) {
+					nodeRacks.Add(value, node.Name)
+				}
+				if !topologyMap.Contains(label, value) {
+					topologyMap.Add(label, value)
 				}
 			}
+			}
```
[the remark this answers is not in the document]
Comment (egafford) [reply pkg/controller/storagecluster/reconcile.go]: I actually don't think the additional indent is needed here. We don't need to iterate over the (label, value) KVPs once for each member of the validTopologyLabelKeys set; we only need to assess each KVP once.

I did mess up moving the closed brace, though, which is stark foolishness and I'm fixing it, so thanks for that.

**Comment by obnoxxx** [comment]
@egafford  writes:
> I actually don't think the additional indent is needed here. We don't need to iterate over the (label, value) KVPs once for each member of the validTopologyLabelKeys set; we only need to assess each KVP once.
> 
> I did mess up moving the closed brace, though, which is stark foolishness and I'm fixing it, so thanks for that.

Ah, that was the other possible explanation. :-)

**Comment by obnoxxx** [comment]
For clarity, and just because I'm always pounding on having as atomic as possible changes in patches, one *could* have moved the change that unindents the if-label-contains-rack-block into a separate preparatory patch in the PR. But let's not get hung up on that right here and now.

File: pkg/controller/storagecluster/reconcile.go [hunk 2]
```diff
@@ -371,10 +371,13 @@ func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageClus
...
-					if !nodeRacks.Contains(value, node.Name) {
-						nodeRacks.Add(value, node.Name)
-					}
+			}
+			if strings.Contains(label, "rack") {
+				if !nodeRacks.Contains(value, node.Name) {
+					nodeRacks.Add(value, node.Name)
+				}
+				if !topologyMap.Contains(label, value) {
+					topologyMap.Add(label, value)
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: I'm confused by the code:

A little above, we have the hunk:
```
if strings.Contains(label, key) {
   if !topologyMap.Contains(label, value) {
      topologyMap.Add(label, value)
   }
} 
```

Now you add:
```
if strings.Contains(label, "rack") {
   if !topologyMap.Contains(label, value) {
      topologyMap.Add(label, value)
   }
}
```
This looks pretty much like a special case. What is really changed in the resulting topology? :thinking:

File: pkg/controller/storagecluster/reconcile.go [hunk 3]
```diff
@@ -671,11 +674,14 @@ func determineFailureDomain(topologyMap *ocsv1.NodeTopologyMap) string {
...
 			if len(labelValues) >= 3 {
 				failureDomain = "zone"
-				break
 			}
 		}
+		// If we have already assigned a rack-based topology to the cluster,
+		// this must be maintained.
+		if strings.Contains(label, "rack") {
+			return "rack"
+		}
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: @egafford, somehow, when discussing this change, I had imagined that we had somehow recorded the failure domain setting in the resulting storage cluster and could somehow retrieve that. But I see what your approach is: I think if I understand your change correctly, you are saying if we see at least one node with a "rack" label, then we have started with rack once, and we are sticking with rack. I see how that should work.

 But it's also has potential problems, introduces a slight risk for bugs for future changes: We currently decide to put rack labels only if we don't do zone. But what if later, we decide to always put rack lables? Or what if we want to define a way to move from zone to rack (and back again?) So the current approach is simple enough, but we need to keep in mind for future changes that this could would need to change if we ever decide to always add rack labels etc. 

Given the simplicity of the patch, I agree that we should probably not do any more complicated approach right now. Possibly note in a comment that it extrapolates from possibly still incomplete label data (`ensureNodeRacks` is invoked later), and assumes knowledge of how the other code works. But maybe I'm just overly paranoid... :wink:

File: pkg/controller/storagecluster/reconcile.go (hunk 2 above)
Comment (obnoxxx) [reply pkg/controller/storagecluster/reconcile.go]: So I think, being a special case, this does not change the resulting topology map at all, and can be omitted. (making the patch even simpler)

File: pkg/controller/storagecluster/reconcile.go (hunk 3 above)
Comment (egafford) [reply pkg/controller/storagecluster/reconcile.go]: No, I see your point. Happily, I think we can absolutely introduce a path to always put rack labels on from here, and racks are inherently zoned, so I don't think we're boxing ourselves in. If in a future release we decide that each zone always consists of racks of n machines, that's easy enough to implement if we're already using zone as the failure domain.

I also agree that a fuller refactor of this code for clarity might help matters (as there absolutely are some assumptions built into my approach to preserve method signatures), but I think that's definitely a between-releases task. For release stabilization, I try to go minimalist. :)

File: pkg/controller/storagecluster/reconcile.go (hunk 2 above)
Comment (obnoxxx) [reply pkg/controller/storagecluster/reconcile.go]: Just confirmed locally that the test passes without this hunk.

**Comment by egafford** [comment]
Sure; I can make this change if we are 100% certain that the existing topologyMap will *always* contain the rack labels assigned to all of its nodes, without fail. I added that logic to guard against a potential case in which a mismatch existed: since determineFailureDomain takes one topologyMap as its argument, it occurred to me that if rack information was somehow present in a node but absent from the cluster's topologyMap, we would miss that information and could assign a failure domain of zone incorrectly. If we believe that this case is formally impossible, though, I'm happy to take it out.

**Comment by jarrpa** [comment]
What about just looking to see if the StorageCluster's `status.failureDomain` is set to something? That should be enough to say "don't recalculate the failure domain", right?

**Comment by obnoxxx** [comment]
@jarrpa writes:
> What about just looking to see if the StorageCluster's status.failureDomain is set to something? That should be enough to say "don't recalculate the failure domain", right?

Dang, that was exactly what I meant when I wrote "I had imagined that we had somehow recorded the failure domain setting in the resulting storage cluster and could somehow retrieve that" ... :-)

That will be the more robust solution, imho. And it will be as simple.

**Comment by obnoxxx** [comment]
@egafford writes:
> Sure; I can make this change if we are 100% certain that the existing topologyMap will always contain the rack labels assigned to all of its nodes, without fail.

I don't think this is the case, and I don't think the patch in `determineFailureDomain()` would require it: It will return "rack" if we have at least one occurrence of a rack label.

> I added that logic to guard against a potential case in which a mismatch existed: since determineFailureDomain takes one topologyMap as its argument, it occurred to me that if rack information was somehow present in a node but absent from the cluster's topologyMap, we would miss that information and could assign a failure domain of zone incorrectly. If we believe that this case is formally impossible, though, I'm happy to take it out.

I think it is, but I also think we should probably consider @jarrpa 's more simple and possibly more robust approach.

**Comment by egafford** [comment]
The difficulty with @jarrpa's suggestion is that determineFailureDomain is called from different scopes. I could change the signature of determineFailureDomain to also (or only) take a StorageCluster, so that the FailureDomain could be assessed. However, this method is also called from newStorageClassDeviceSets and a different point in the main Reconcile loop. In newStorageClassDeviceSets, we don't have a ref to the entire StorageCluster.

I can absolutely make the change to passing the StorageCluster to newStorageClassDeviceSets so that we can assess via this method. Is this preferable to the current code?

File: pkg/controller/storagecluster/reconcile.go [hunk 4]
```diff
@@ -371,17 +371,17 @@ func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageClus
...
 					}
 				}
-				if strings.Contains(label, "rack") {
-					if !nodeRacks.Contains(value, node.Name) {
-						nodeRacks.Add(value, node.Name)
-					}
+			}
+			if strings.Contains(label, "rack") {
+				if !nodeRacks.Contains(value, node.Name) {
+					nodeRacks.Add(value, node.Name)
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: This could/should still have been a separate commit, as it is a code cleanup/improvement not needed for actual fix that is mentioned in the commit message.

File: pkg/controller/storagecluster/reconcile.go [hunk 5]
```diff
@@ -664,18 +664,19 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1.StorageCluster, re
...
+		return sc.Status.FailureDomain
+	}
+	topologyMap := sc.Status.NodeTopologies
 	failureDomain := "rack"
-
 	for label, labelValues := range topologyMap.Labels {
 		if strings.Contains(label, "zone") {
 			if len(labelValues) >= 3 {
 				failureDomain = "zone"
-				break
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: Why did you remove the break?

File: pkg/controller/storagecluster/reconcile.go [hunk 6]
```diff
@@ -664,18 +664,19 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1.StorageCluster, re
...
 			}
 		}
 	}
-
```
Comment (obnoxxx) [inline pkg/controller/storagecluster/reconcile.go]: Just a taste thing (and I know that e.g. @jarrpa is not as focused on these as I am): I personally do like some new lines to group the code more for my eyes, but more importantly, I don't like changes that are not related to the actual change to be sneaked in, even if they are just whitespace changes.

So my rule of thumb would be: if you are changing this context anyway, feel free to do whitespace or formatting changes in the same commit, but if there are no changes anyway in this context the commit, keep the diff smaller. If you still want to do the whitespace or formatting changes, do them as separate commits maybe even in the same PR.
